audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+	// so that we can read it without a lock. We use a lock to write to the error. This is because
+	// we want to ensure that the connection is only closed once. Reads of err are fast, setting an
+	// error is slower, but only happens once.
+
+	// err is the last error that occurred. This is used to determine if the client is in an
+	// unrecoverable state.
+	err atomic.Pointer[error]
+	// setErrMu is used to to lock write changes to err. Reads can simply use atomic.Value.Load().
+	setErrMu sync.Mutex
+
+	closeOnceMu sync.Mutex
+	// closeOnce is used to ensure that the client is only closed once.
+	closeOnce sync.Once
+
+	// log is the logger for the audit client. If not set, the default logger is used.
+	log *slog.Logger
+
+	// heartbeat is the message for the heartbeat.
+	heartbeat msgs.Msg
+	// heartbeatInterval is the interval for sending heartbeats. A value of
+	// 0 means no heartbeats are sent.
+	heartbeatInterval time.Duration
+}
+
+// Settings represents the settings for the audit client. These are used to configure how the
+// client behaves. The zero value is valid and will use the default values. Negative values
+// are invalid and will be set to the default value.
+type Settings struct {
+	// QueueSize is the maximum number of audit records that can be queued.
+	// Defaults to MaxQueueSize. This queue is the queue not only for sending records
+	// but Records that fail to send for any reason but validation and context timeouts
+	// will be requeued here. When this is full, the record is dropped.
+	QueueSize int
+}
+
+// defaults sets the default values for the settings. It returns a copy of settings to avoid an allocation.
+func (s Settings) defaults() Settings {
+	if s.QueueSize < 1 {
+		s.QueueSize = MaxQueueSize
+	}
+	return s
+}
+
+// Option represents an option for the audit client.
+type Option func(c *Client) error
+
+// WithSettings sets the settings for the audit client. If not set, the default values are used.
+func WithSettings(s Settings) Option {
+	return func(c *Client) error {
+		c.settings = s.defaults()
+		return nil
+	}
+}
+
+// WithLogger sets the logger for the audit client. If not set, the default logger is used.
+// This is for the few log messages that can occur, such as message drops. As a note, slog.Logger
+// is not an interface. But loggers are created with a handler which tell the logger where to
+// send the log messages. You can use that to convert a slog.Logger to a different logger if needed.
+// Also, check with your logging package, as it may have a way to send slog.Logger messages to it.
+func WithLogger(l *slog.Logger) Option {
+	return func(c *Client) error {
+		c.log = l
+		return nil
+	}
+}
+
+// New returns a new audit client.
+func New(c conn.Audit, options ...Option) (client *Client, err error) {
+	if c == nil {
+		return nil, Error{Err: fmt.Errorf("cannot pass a conn.Audit that is nil"), Category: ErrConnection}
+	}
+
+	cli := &Client{
+		settings:          Settings{}.defaults(),
```

Comment:
minor: Why not just call `Settings{}.defaults()` after the range of options is applied, and drop this call to it and the call in `WithSettings`?

Reply:
This basically:
```
	cli := &Client{
		settings:          Settings{},
		...
	}
	...
	for _, o := range options {
		if err := o(cli); err != nil {
			return nil, err
		}
	}
	cli.settings = cli.settings.default() // Apply any defaults that the user has not overridden
```

Reply:
TBH I'm not sold this is really all that measurably better than what you have, the main advantage to it in my mind is that there's only 1 place defaults are applied as opposed to two like you've got it now.

Reply:
Well, your version is assuming that everything can be handled in defaults() as I am doing with queue size.  By setting defaults early, anything done with options is changing the defaults to a new overriden state.

So if QueueSize defaults to 2048, after the options are applied I know any deviation from 2048 was set by the caller.

But what if 0 is a valid value (unlike my QueueSize), Did the user change it?  How do I know if I should default it to 2048?

If you setup the defaults at the beginning (either by doing it with a composite literal or a default() method), after the options are applied, you don't have to make guess work.

Reply:
Color me convinced.